### PR TITLE
SDT-20 Rebuild DB build scripts

### PR DESCRIPTION
**Pull request by markdathornehmcts**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/SDT-20

### Change description ###

Rebuild DB build scripts

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[*] No
```

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -196,20 +196,20 @@ allprojects {
...
-      'services/build/libs/civil-sdt-services-0.0.1-SNAPSHOT.jar',
-      'services-api/build/libs/civil-sdt-services-api-0.0.1-SNAPSHOT.jar',
-      'transformers/build/libs/civil-sdt-transformers-0.0.1-SNAPSHOT.jar',
-      'utils/build/libs/civil-sdt-utils-0.0.1-SNAPSHOT.jar',
-      'validators/build/libs/civil-sdt-validators-0.0.1-SNAPSHOT.jar')
+      "consumers/build/libs/civil-sdt-consumers-${version}.jar",
+      "dao/build/libs/civil-sdt-dao-${version}.jar",
+      "dao-api/build/libs/civil-sdt-dao-api-${version}.jar",
+      "domain/build/libs/civil-sdt-domain-${version}.jar",
+      "handlers/build/libs/civil-sdt-handlers-${version}.jar",
```
Comment (kiran-yenigala-hmcts) [inline build.gradle:203]: why this change is here

File: build.gradle [hunk 2]
```diff
@@ -567,15 +567,19 @@ idea {
...
 bootJar {
-
   manifest {
     attributes 'Implementation-Title': project.name,
             'Implementation-Version': project.version
   }
+  duplicatesStrategy = DuplicatesStrategy.EXCLUDE
   archiveName = 'civil-sdt.jar'
+  from {
+    subprojects.collect { it.sourceSets.main.output }
```
Comment (kiran-yenigala-hmcts) [inline build.gradle:577]: same as above comment

File: src/main/resources/db/migration/V0001__init.sql [hunk 3]
```diff
@@ -0,0 +1,481 @@
...
+,'The Target Application does not send back an acknowledgement response within the expected period, or returns a response indicating that there was an error with the transmission.');
+INSERT INTO "SDT_OWNER"."ERROR_MESSAGES" ("ERROR_MESSAGE_ID","ERROR_CODE","ERROR_TEXT","ERROR_DESCRIPTION")
+VALUES (14,'CUST_XML_ERR','Individual Request format could not be processed by the Target Application. Please check the data and resubmit the request, or contact {0} for assistance.'
+,'Client data has caused SOAP Fault error and rejected the request.');
+
+
+
+
+
+
```
Comment (kiran-yenigala-hmcts) [inline src/main/resources/db/migration/V0001__init.sql:474]: can we please delete empty lines at the end

**Review by kiran-yenigala-hmcts — CHANGES_REQUESTED** [review]
Minor comments added

File: build.gradle (hunk 1 above)
Comment (paulridout) [reply build.gradle:203]: I think these were originally part of #28 then subsequently changed.

File: src/main/resources/db/migration/V0001__init.sql [hunk 4]
```diff
@@ -0,0 +1,481 @@
...
+-------------------------------------------------------------
+-- Drop each of the users used by SD
+--------------------------------------------------------------
+---- non schema owners first
+--DROP USER IF EXISTS "SDT_USER";
+--DROP USER IF EXISTS "SDT_BATCH_USER";
+---- now schema owners
+--DROP USER IF EXISTS "SDT_OWNER";
+
```
Comment (paulridout) [inline src/main/resources/db/migration/V0001__init.sql:16]: Any reason why these lines have been commented out?  The comment on line 9 seems to be missing the "T" from the end of SDT.

File: src/main/resources/db/migration/V0001__init.sql [hunk 5]
```diff
@@ -0,0 +1,481 @@
...
+------------------------------------------------
+-- Create tables for SDT_OWNER
+------------------------------------------------
+CREATE TABLE "SDT_OWNER"."BULK_CUSTOMERS"
+   ("BULK_CUSTOMER_ID" NUMERIC NOT NULL, "SDT_CUSTOMER_ID" NUMERIC(8),
+	"VERSION_NUMBER" NUMERIC DEFAULT 0 ) ;
+
+CREATE TABLE "SDT_OWNER"."BULK_CUSTOMER_APPLICATIONS"
+   ("BULK_CUSTOMER_APPLICATIONS_ID" NUMERIC NOT NULL,
+	"BULK_CUSTOMER_ID" NUMERIC,	"TARGET_APPLICATION_ID" NUMERIC,
```
Comment (paulridout) [inline src/main/resources/db/migration/V0001__init.sql:44]: Replace the tab after the first comma with a space.  There's quite a few more instances of this throughout the file.

File: src/main/resources/db/migration/V0001__init.sql [hunk 6]
```diff
@@ -0,0 +1,481 @@
...
+CREATE USER "SDT_USER";
+GRANT POSTGRES TO "SDT_USER";
+CREATE USER "SDT_BATCH_USER";
+GRANT POSTGRES TO "SDT_BATCH_USER";
+
+------------------------------------------------
+-- Create tables for SDT_OWNER
+------------------------------------------------
+CREATE TABLE "SDT_OWNER"."BULK_CUSTOMERS"
+   ("BULK_CUSTOMER_ID" NUMERIC NOT NULL, "SDT_CUSTOMER_ID" NUMERIC(8),
```
Comment (paulridout) [inline src/main/resources/db/migration/V0001__init.sql:39]: Do the primary key columns need to be declared as "not null" within the table declarations?  Just wondering if the primary key declarations later on automatically set the columns to be not null.

File: src/main/resources/db/migration/V0001__init.sql [hunk 7]
```diff
@@ -0,0 +1,481 @@
...
+CREATE INDEX "SR_SERVICE_TYPE_I" ON "SDT_OWNER"."SERVICE_ROUTINGS" ("SERVICE_TYPE_ID");
+CREATE INDEX "SR_TARGET_APPLICATION_I" ON "SDT_OWNER"."SERVICE_ROUTINGS" ("TARGET_APPLICATION_ID");
+CREATE INDEX "ST_SERVICE_TYPE_NAME" ON "SDT_OWNER"."SERVICE_TYPES" ("SERVICE_TYPE_NAME");
+CREATE UNIQUE INDEX "TARGET_APPLICATIONS_PK" ON "SDT_OWNER"."TARGET_APPLICATIONS" ("TARGET_APPLICATION_ID");
+CREATE INDEX "TA_TARGET_APPLICATION_NAME" ON "SDT_OWNER"."TARGET_APPLICATIONS" ("TARGET_APPLICATION_NAME");
+
+------------------------------------------------
+-- Create Primary Keys for SDT_OWNER
+------------------------------------------------
+ALTER TABLE "SDT_OWNER"."BULK_CUSTOMERS" ADD CONSTRAINT "BULK_CUSTOMERS_PK" PRIMARY KEY USING INDEX "BULK_CUSTOMERS_PK";
```
Comment (paulridout) [inline src/main/resources/db/migration/V0001__init.sql:138]: Do indexes need to be prefixed with owner (e.g. "SDT_OWNER"."BULK_CUSTOMERS_PK")?

File: src/main/resources/db/migration/V0001__init.sql [hunk 8]
```diff
@@ -0,0 +1,481 @@
...
+ALTER TABLE "SDT_OWNER"."INDIVIDUAL_REQUESTS" ADD CONSTRAINT "INDIVIDUAL_REQUESTS_PK" PRIMARY KEY USING INDEX "INDIVIDUAL_REQUESTS_PK";
+ALTER TABLE "SDT_OWNER"."SERVICE_REQUESTS" ADD CONSTRAINT "SERVICE_REQUESTS_PK" PRIMARY KEY USING INDEX "SERVICE_REQUESTS_PK";
+ALTER TABLE "SDT_OWNER"."SERVICE_ROUTINGS" ADD CONSTRAINT "SERVICE_ROUTINGS_PK" PRIMARY KEY USING INDEX "SERVICE_ROUTINGS_PK";
+ALTER TABLE "SDT_OWNER"."SERVICE_TYPES" ADD CONSTRAINT "SERVICE_TYPES_PK" PRIMARY KEY USING INDEX "SERVICE_TYPES_PK";
+ALTER TABLE "SDT_OWNER"."TARGET_APPLICATIONS" ADD CONSTRAINT "TARGET_APPLICATIONS_PK" PRIMARY KEY USING INDEX "TARGET_APPLICATIONS_PK";
+
+------------------------------------------------
+-- Create Check Constraints for SDT_OWNER
+------------------------------------------------
+-- capitalised names reqd?
```
Comment (paulridout) [inline src/main/resources/db/migration/V0001__init.sql:153]: Don't forget to remove this comment when done.

File: src/main/resources/db/migration/V0001__init.sql [hunk 9]
```diff
@@ -0,0 +1,481 @@
...
+CREATE SEQUENCE  "SDT_OWNER"."GLB_PAR_SEQ"  MINVALUE 1 MAXVALUE 999999999 INCREMENT BY 1 START WITH 1 CYCLE;
+CREATE SEQUENCE  "SDT_OWNER"."IND_REQ_SEQ"  MINVALUE 1 MAXVALUE 999999999999999999 INCREMENT BY 1 START WITH 1 CACHE 2000 NO CYCLE;
+CREATE SEQUENCE  "SDT_OWNER"."SDT_REF_SEQ"  MINVALUE 1 MAXVALUE 999999999 INCREMENT BY 1 START WITH 1 CYCLE;
+CREATE SEQUENCE  "SDT_OWNER"."SER_ROU_SEQ"  MINVALUE 1 MAXVALUE 999999999 INCREMENT BY 1 START WITH 1 CYCLE;
+CREATE SEQUENCE  "SDT_OWNER"."SER_TYP_SEQ"  MINVALUE 1 MAXVALUE 999999999 INCREMENT BY 1 START WITH 1 CYCLE;
+CREATE SEQUENCE  "SDT_OWNER"."SRV_REQ_SEQ"  MINVALUE 1 MAXVALUE 999999999999999999 INCREMENT BY 1 START WITH 1 NO CYCLE;
+CREATE SEQUENCE  "SDT_OWNER"."TAR_APP_SEQ"  MINVALUE 1 MAXVALUE 999999999 INCREMENT BY 1 START WITH 1 CYCLE;
+
+------------------------------------------------
+-- TOD: Grants! Tablespaces Synonym for SDT_OWNER
```
Comment (paulridout) [inline src/main/resources/db/migration/V0001__init.sql:242]: Should "TOD" be "TODO"?

File: src/main/resources/db/migration/V0001__init.sql [hunk 10]
```diff
@@ -0,0 +1,481 @@
...
+--GRANT UNLIMITED TABLESPACE TO "SDT_OWNER";
+--GRANT CREATE SYNONYM TO "SDT_USER";
+--GRANT CREATE SYNONYM TO "SDT_BATCH_USER";
+-----------------
+--- Sys Grants
+-----------------
+--GRANT EXECUTE ON "SYS"."UTL_FILE" TO "SDT_OWNER";
+
+------------------------------------------------
+-- Create Views (not Synonyms!) for SDT_USER
```
Comment (paulridout) [inline src/main/resources/db/migration/V0001__init.sql:328]: Might be better to say creating views to replace synonyms.

File: src/main/resources/db/migration/V0001__init.sql [hunk 11]
```diff
@@ -0,0 +1,481 @@
...
+------------------------------------------------
+CREATE OR REPLACE VIEW "SDT_USER"."BULK_CUSTOMERS" AS SELECT * FROM "SDT_OWNER"."BULK_CUSTOMERS";
+CREATE OR REPLACE VIEW "SDT_USER"."BULK_CUSTOMER_APPLICATIONS" AS SELECT * FROM "SDT_OWNER"."BULK_CUSTOMER_APPLICATIONS";
+CREATE OR REPLACE VIEW "SDT_USER"."BULK_CUST_APP_SEQ" AS SELECT * FROM "SDT_OWNER"."BULK_CUST_APP_SEQ";
```
Comment (paulridout) [inline src/main/resources/db/migration/V0001__init.sql:332]: Does this work for sequences?

File: src/main/resources/db/migration/V0001__init.sql [hunk 12]
```diff
@@ -0,0 +1,481 @@
...
+CREATE OR REPLACE VIEW "SDT_USER"."BULK_SUB_SEQ" AS SELECT * FROM "SDT_OWNER"."BULK_SUB_SEQ";
+CREATE OR REPLACE VIEW "SDT_USER"."ERROR_LOGS" AS SELECT * FROM "SDT_OWNER"."ERROR_LOGS";
+CREATE OR REPLACE VIEW "SDT_USER"."ERROR_MESSAGES" AS SELECT * FROM "SDT_OWNER"."ERROR_MESSAGES";
+CREATE OR REPLACE VIEW "SDT_USER"."ERR_LOG_SEQ" AS SELECT * FROM "SDT_OWNER"."ERR_LOG_SEQ";
+CREATE OR REPLACE VIEW "SDT_USER"."ERR_MESG_SEQ" AS SELECT * FROM "SDT_OWNER"."ERR_MESG_SEQ";
+CREATE OR REPLACE VIEW "SDT_USER"."GLB_PAR_SEQ" AS SELECT * FROM "SDT_OWNER"."GLB_PAR_SEQ";
+CREATE OR REPLACE VIEW "SDT_USER"."GLOBAL_PARAMETERS" AS SELECT * FROM "SDT_OWNER"."GLOBAL_PARAMETERS";
+CREATE OR REPLACE VIEW "SDT_USER"."INDIVIDUAL_REQUESTS" AS SELECT * FROM "SDT_OWNER"."INDIVIDUAL_REQUESTS";
+CREATE OR REPLACE VIEW "SDT_USER"."IND_REQ_SEQ" AS SELECT * FROM "SDT_OWNER"."IND_REQ_SEQ";
+--CREATE OR REPLACE VIEW "SDT_USER"."MESSAGE_LOGS" AS SELECT * FROM "SDT_OWNER"."MESSAGE_LOGS";
```
Comment (paulridout) [inline src/main/resources/db/migration/V0001__init.sql:344]: message_logs may no longer exist.  Can remove this if that's correct.

File: src/main/resources/db/migration/V0001__init.sql [hunk 13]
```diff
@@ -0,0 +1,481 @@
...
+CREATE OR REPLACE VIEW "SDT_USER"."TAR_APP_SEQ" AS SELECT * FROM "SDT_OWNER"."TAR_APP_SEQ";
+
+------------------------------------------------
+-- Create Views (Not Synonyms!) for SDT_BATCH_USER
+------------------------------------------------
+CREATE OR REPLACE VIEW "SDT_BATCH_USER"."BULK_SUBMISSIONS" AS SELECT * FROM "SDT_OWNER"."BULK_SUBMISSIONS";
+CREATE OR REPLACE VIEW "SDT_BATCH_USER"."ERROR_LOGS" AS SELECT * FROM "SDT_OWNER"."ERROR_LOGS";
+CREATE OR REPLACE VIEW "SDT_BATCH_USER"."GLOBAL_PARAMETERS" AS SELECT * FROM "SDT_OWNER"."GLOBAL_PARAMETERS";
+CREATE OR REPLACE VIEW "SDT_BATCH_USER"."INDIVIDUAL_REQUESTS" AS SELECT * FROM "SDT_OWNER"."INDIVIDUAL_REQUESTS";
+--CREATE OR REPLACE VIEW "SDT_BATCH_USER"."PURGE" AS SELECT * FROM "SDT_OWNER"."PURGE";
```
Comment (paulridout) [inline src/main/resources/db/migration/V0001__init.sql:362]: Remember to remove this commented out line.

File: src/main/resources/db/migration/V0001__init.sql (hunk 4 above)
Comment (markdathornehmcts) [reply src/main/resources/db/migration/V0001__init.sql:16]: Because when run from Flyway, it's the first script and will be a new database. So only while continually re-testing, this is needed.

File: build.gradle (hunk 1 above)
Comment (markdathornehmcts) [reply build.gradle:203]: Either caused from timing of the branch creation (or merges since)

File: src/main/resources/db/migration/V0001__init.sql (hunk 7 above)
Comment (markdathornehmcts) [reply src/main/resources/db/migration/V0001__init.sql:138]: No. Doesn't like that syntax.

File: src/main/resources/db/migration/V0001__init.sql (hunk 8 above)
Comment (markdathornehmcts) [reply src/main/resources/db/migration/V0001__init.sql:153]: Done

File: src/main/resources/db/migration/V0001__init.sql (hunk 11 above)
Comment (markdathornehmcts) [reply src/main/resources/db/migration/V0001__init.sql:332]: No

File: src/main/resources/db/migration/V0001__init.sql [hunk 14]
```diff
@@ -0,0 +1,481 @@
...
+GRANT DELETE ON "SDT_OWNER"."BULK_SUBMISSIONS" TO "SDT_BATCH_USER";
+GRANT SELECT ON "SDT_OWNER"."BULK_SUBMISSIONS" TO "SDT_BATCH_USER";
+GRANT DELETE ON "SDT_OWNER"."ERROR_LOGS" TO "SDT_BATCH_USER";
+GRANT SELECT ON "SDT_OWNER"."ERROR_LOGS" TO "SDT_BATCH_USER";
+GRANT SELECT ON "SDT_OWNER"."GLOBAL_PARAMETERS" TO "SDT_BATCH_USER";
+GRANT DELETE ON "SDT_OWNER"."INDIVIDUAL_REQUESTS" TO "SDT_BATCH_USER";
+GRANT SELECT ON "SDT_OWNER"."INDIVIDUAL_REQUESTS" TO "SDT_BATCH_USER";
+GRANT DELETE ON "SDT_OWNER"."SERVICE_REQUESTS" TO "SDT_BATCH_USER";
+GRANT SELECT ON "SDT_OWNER"."SERVICE_REQUESTS" TO "SDT_BATCH_USER";
+--GRANT EXECUTE ON "SDT_OWNER"."PURGE" TO "SDT_BATCH_USER";
```
[the remark this answers is not in the document]
Comment (markdathornehmcts) [reply src/main/resources/db/migration/V0001__init.sql:241]: Keeping this until we're sure it's no needed - answer may be in SDT-19

File: src/main/resources/db/migration/V0001__init.sql (hunk 13 above)
Comment (markdathornehmcts) [reply src/main/resources/db/migration/V0001__init.sql:337]: May be reintroduced as part of SDT-19 so keeping, for now,

File: src/main/resources/db/migration/V0001__init.sql (hunk 5 above)
Comment (markdathornehmcts) [reply src/main/resources/db/migration/V0001__init.sql]: Not necessary but done.

File: src/main/resources/db/migration/V0001__init.sql (hunk 6 above)
Comment (markdathornehmcts) [reply src/main/resources/db/migration/V0001__init.sql]: Unclear. No change.

File: src/main/resources/db/migration/V0001__init.sql (hunk 5 above)
Comment (paulridout) [reply src/main/resources/db/migration/V0001__init.sql]: It is a cosmetic change, but it does ensure that a consistent spacing scheme is used throughout the file.

There are still seven tabs left:
Line 42 - after comma at the end of '"BULK_CUSTOMER_ID" NUMERIC,'
Line 43 - after comma at the end of '"CUSTOMER_APPLICATION_ID" VARCHAR(32),'
Line 46 - after first two commas (they look like spaces but are tabs)
Line 49 - after comma at the end of '"COMPLETED_DATE" TIMESTAMP(6),'
Line 50 - after comma at the end of 'VERSION_NUMBER" NUMERIC DEFAULT 0,' (looks like a space but is a tab)
Line 80 - after comma at end of '"SERVICE_TYPE_ID" NUMERIC,' (looks like a space but is a tab)

File: src/main/resources/db/migration/V0001__init.sql (hunk 11 above)
Comment (paulridout) [reply src/main/resources/db/migration/V0001__init.sql:308]: If we can't define views for sequences are we going to make any changes so that they can be accessed?

File: src/main/resources/db/migration/V0001__init.sql (hunk 12 above)
Comment (paulridout) [reply src/main/resources/db/migration/V0001__init.sql]: Can't see any other references to MESSAGE_LOGS in the code so it appears this is redundant and can be removed.

File: src/main/resources/db/migration/V0001__init.sql (hunk 11 above)
Comment (markdathornehmcts) [reply src/main/resources/db/migration/V0001__init.sql:308]: Ah! I thought you meant the OR REPLACE

Comment (markdathornehmcts) [reply src/main/resources/db/migration/V0001__init.sql:308]: We just followed the script and create Views in place of Synonyms where it was declared. How do we know which sequences to make accessible to the two users?

File: src/main/resources/db/migration/V0001__init.sql (hunk 9 above)
Comment (markdathornehmcts) [reply src/main/resources/db/migration/V0001__init.sql]: Line obliterated

**Review by kiran-yenigala-hmcts — APPROVED** [review]

File: src/main/resources/db/migration/V0001__init.sql [hunk 15]
```diff
@@ -0,0 +1,482 @@
...
+ALTER ROLE sdt_batch_user SET search_path TO sdt_batch_user, sdt_owner;
+
+GRANT USAGE ON SCHEMA sdt_owner TO sdt_user;
+GRANT USAGE ON SCHEMA sdt_owner TO sdt_batch_user;
+
+GRANT ALL PRIVILEGES ON ALL SEQUENCES IN SCHEMA sdt_owner TO sdt_user;
+GRANT ALL PRIVILEGES ON ALL SEQUENCES IN SCHEMA sdt_owner TO sdt_batch_user;
+
+GRANT ALL PRIVILEGES ON ALL TABLES IN SCHEMA sdt_owner TO sdt_user;
+GRANT ALL PRIVILEGES ON ALL TABLES IN SCHEMA sdt_owner TO sdt_batch_user;
```
Comment (paulridout) [inline src/main/resources/db/migration/V0001__init.sql:325]: These GRANT ALL statements and the following Object Grants section are mutually exclusive.  We only want to grant specific access to specific objects for each user - which is covered by the Object Grants section - so it doesn't look like the GRANT ALL statements are needed.

File: src/main/resources/db/migration/V0001__init.sql [hunk 16]
```diff
@@ -0,0 +1,482 @@
...
+CREATE OR REPLACE VIEW sdt_batch_user.service_requests AS SELECT * FROM sdt_owner.service_requests;
+
+
+------------------------------------------------
+-- Create Grants for SDT_OWNER
+------------------------------------------------
+
+ALTER ROLE sdt_owner SET search_path TO sdt_owner;
+ALTER ROLE sdt_user SET search_path TO sdt_user, sdt_owner;
+ALTER ROLE sdt_batch_user SET search_path TO sdt_batch_user, sdt_owner;
```
Comment (paulridout) [inline src/main/resources/db/migration/V0001__init.sql:316]: Do we still need the views with these search_path settings?  For example, can you execute SELECT * FROM sdt_user.bulk_customers as sdt_user without the sdt_user.bulk_customers view defined?  If the views aren't needed then let's remove them.

Comment (markdathornehmcts) [reply src/main/resources/db/migration/V0001__init.sql:277]: Proven so removing views.

File: src/main/resources/db/migration/V0001__init.sql (hunk 15 above)
Comment (markdathornehmcts) [reply src/main/resources/db/migration/V0001__init.sql]: Done

File: src/main/resources/db/migration/V0001__init.sql (hunk 5 above)
Comment (paulridout) [reply src/main/resources/db/migration/V0001__init.sql]: The CREATE table statements have been changed to have each column on a new line so these comments no longer apply.

**Review by kiran-yenigala-hmcts — APPROVED** [review]
